src/lib.rs

```diff
@@ -51,80 +51,107 @@ extern "C" {
+            BAD_GPIO => Err("Bad gpio".to_string()),
+            BAD_PUD => Err("Bad pud".to_string()),
+            _ => Err(DEFAULT_ERROR.to_string())
+        }
     }
-}
 
-/// Gets the GPIO mode.
-pub fn get_mode(gpio: u32) -> GpioResponse {
-    let response = unsafe { gpioGetMode(gpio) };
-    match response {
-        BAD_GPIO => Err("Bad gpio".to_string()),
-        _ => Ok(response as u32),
+    /// Reads the GPIO level, on or off.
+    pub fn read(&self, gpio: u32) -> GpioResponse {
+        match unsafe { gpioRead(gpio) } {
+            1 => Ok(1),
+            0 => Ok(0),
+            BAD_GPIO => Err("Bad gpio".to_string()),
+            _ => Err(DEFAULT_ERROR.to_string()),
+        }
     }
-}
 
-/// Sets or clears resistor pull ups or downs on the GPIO.
-pub fn set_pull_up_down(gpio: u32, pud: Pud) -> GpioResult {
-    match unsafe { gpioSetPullUpDown(gpio, pud as u32) } {
-        OK => Ok(()),
-        BAD_GPIO => Err("Bad gpio".to_string()),
-        BAD_PUD => Err("Bad pud".to_string()),
-        _ => Err(DEFAULT_ERROR.to_string())
+    /// Sets the GPIO level, on or off.
+    /// If PWM or servo pulses are active on the GPIO they are switched off.
+    pub fn write(&self, gpio: u32, level: Level) -> GpioResult {
+        match unsafe { gpioWrite(gpio, level as u32) } {
+            OK => Ok(()),
+            BAD_GPIO => Err("Bad gpio".to_string()),
+            BAD_LEVEL => Err("Bad level".to_string()),
+            _ => Err(DEFAULT_ERROR.to_string()),
+        }
     }
-}
 
-/// Reads the GPIO level, on or off.
-pub fn read(gpio: u32) -> GpioResponse {
-    match unsafe { gpioRead(gpio) } {
-        1 => Ok(1),
-        0 => Ok(0),
-        BAD_GPIO => Err("Bad gpio".to_string()),
-        _ => Err(DEFAULT_ERROR.to_string()),
+    /// Delays for at least the number of microseconds specified by microseconds.
+    pub fn delay(&self, microseconds: u32) -> u32 {
+        unsafe { gpioDelay(microseconds) }
     }
 }
 
-/// Sets the GPIO level, on or off.
-/// If PWM or servo pulses are active on the GPIO they are switched off.
-pub fn write(gpio: u32, level: Level) -> GpioResult {
-    match unsafe { gpioWrite(gpio, level as u32) } {
-        OK => Ok(()),
-        BAD_GPIO => Err("Bad gpio".to_string()),
-        BAD_LEVEL => Err("Bad level".to_string()),
-        _ => Err(DEFAULT_ERROR.to_string()),
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn set_gpio_mode_before_init_should_fail() {
+        let response = unsafe { gpioSetMode(8, INPUT as u32) };
+
+        assert_ne!(0, response);
     }
-}
 
-/// Delays for at least the number of microseconds specified by microseconds.
-pub fn delay(microseconds: u32) -> u32 {
-    unsafe { gpioDelay(microseconds) }
+   #[test]
+   fn set_mode_after_drop_should_fail() {
+       {
+           let pigpio = Pigpio::new().unwrap();
+           pigpio.set_mode(8, INPUT).unwrap();
+       }
+
+       let response = unsafe { gpioSetMode(8, INPUT as u32) };
```

Comment:
Hmmm. This is testing using the "private" c api. I guess I'm most interested in making sure the gpioTerminate() is always called when it goes out of scope aka the destructor is called. Is there a way to do this? Maybe mock the gpioTerminate call? If there isn't a good way to do this then maybe this test is sufficient.

Do these tests only run on the pi?

Can you write one more test to demonstrate what happens when to Pigpio::new() calls are made in the same scope? Second should fail right?

Thanks for the effort here!

Reply:
Yes, this is a big problem. I don't think having a drop trait guarantees that it will be called. I don't know if the Rust developers have changed it, but in our experience drop isn't called on a panic. This is a problem because after a few times closing without terminating I would get the error "initMboxBlock: init mbox zaps failed" since the memory from the previous sessions wasn't deallocated. The only fix I found was restarting the Pi.

Reply:
@martindeegan I don't think this is a regression though in that case. Both before, and with this change a panic causes that. The drop is nice in that you can't forget to close it when pgpio goes out of scope.

Reply:
The documentation for Drops say it is called on a panic: https://doc.rust-lang.org/std/ops/trait.Drop.html#panics

How can Mock Something in a rust test? I have searched and found no good answer. Have you a example?

I had run this on the pi. But with mocking, we can get this working as a real unit test. But this test only if the documentation of the Drop trait is correct done by the rust compiler.

Reply:
Yea... You could implement the underlying dynamically linked c library to return whatever you want but that would be problematic when running on the pi as I'm not sure how to specify which version gets linked.  I think you can implement two versions of drop using #[cfg(test)]  and #[cfg(not(test))]. The test version could flip a flag which the test would assert was flipped.

Can you implement the double new() call and make sure the second one fails? Also simulating a panic, if possible would be great.

Reply:
It would be nice though to implement the skeleton of the c lib with some mock functionality in the future for the purpose of running this off the PI.

Reply:
Added a extra test for double new.